README
NAME
    Plack::Middleware::Debug::Redis - Extend Plack::Middleware::Debug with
    Redis panels

VERSION
    version 0.02

SYNOPSIS
        # inside your psgi app
        use Plack::Builder;

        my $app = sub {[
            200,
            [ 'Content-Type' => 'text/html' ],
            [ '<html><body>OK</body></html>' ]
        ]};
        my $redis_host = 'redi.example.com:6379';

        builder {
            mount '/' => builder {
                enable 'Debug',
                    panels => [
                        [ 'Redis::Info', server => $redis_host ],
                        [ 'Redis::Keys', server => $redis_host, db => 3 ],
                    ];
                $app;
            };
        };

DESCRIPTION
    This distribution extends Plack::Middleware::Debug with some Redis
    panels. At the moment, the following panels available:

PANELS
  Redis::Info
    Diplay panel with generic Redis server information which is available by
    the command INFO. See Plack::Middleware::Debug::Redis::Info for
    additional information.

  Redis::Keys
    Diplay panel with keys Redis server information. See
    Plack::Middleware::Debug::Redis::Keys for additional information.

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

SEE ALSO
    Plack::Middleware::Debug::Redis::Info

    Plack::Middleware::Debug::Redis::Keys

    Plack::Middleware::Debug

    Redis

AUTHOR
    Anton Gerasimov, <chim@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Anton Gerasimov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Plack::Middleware::Debug::Redis.

0.02  Mon Jan 14 05:35:26 2013
    - added LICENSE
    - updated build prerequisites
    - in *.t Redis mocked with Test::MockObject

0.01  Tue Jan  8 12:30:26 2013
	- original version


t/lib/FakeRedis.pm
#
# This file is part of Dancer-Plugin-Redis
#
# This software is copyright (c) 2011 by celogeek <me@celogeek.com>.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package FakeRedis;

use strict;
use warnings;
{

    package Redis;
    $INC{'Redis.pm'} = 1;
    our $VERSION = 2;
    our $AUTOLOAD;

    # faked 'INFO'
    my $INFO = {
        'redis_version'     => '0.1.99',
        'db0'               => 'keys=167,expires=145',
        'db1'               => 'keys=75,expires=0',
        'uptime_in_seconds' => '1591647',
        'role'              => 'master',
    };

    # faked 'KEYS'
    my $KEYS = {
        # STRING
        'coy:knows:pseudonoise:codes'       => [ 'string', 9000 ],
        # LIST
        'six:slimy:snails:sailed:silently'  => [ 'list', 35 ],
        # HASH
        'eleven:benevolent:elephants'       => [ 'hash', 17 ],
        # SET
        'two:tried:and:true:tridents'       => [ 'set', 101 ],
        # ZSET
        'tie:twine:to:three:tree:twigs'     => [ 'zset', 66 ],
    };

    sub new {
        my ( $class, %args ) = @_;
        bless \%args => $class;
    }

    sub keys {
        keys %$KEYS;
    }

    sub type {
        exists $KEYS->{$_[1]} ? $KEYS->{$_[1]}->[0] : undef;
    }

    sub info {
        $INFO;
    }

    sub _generic_len {
        exists $KEYS->{$_[1]} ? $KEYS->{$_[1]}->[1] : undef;
    }

    *strlen = *_generic_len;
    *hlen   = *_generic_len;
    *llen   = *_generic_len;
    *scard  = *_generic_len;
    *zcard  = *_generic_len;

    sub AUTOLOAD {
        shift;
        my $name = $AUTOLOAD;
        $name =~ s/.*://;
        return $name, @_;
    }
}

1;


Makefile.PL
use v5.10.1;
use ExtUtils::MakeMaker;

my $meta_merge = {
    META_MERGE => {
        resources => {
            repository => 'https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis',
            bugtracker => 'https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues',
        },
    }
};

my %options = (
    'NAME'          => 'Plack::Middleware::Debug::Redis',
    'VERSION_FROM'  => 'lib/Plack/Middleware/Debug/Redis.pm',
    'AUTHOR'        => 'Anton Gerasimov <chim@cpan.org>',
    'ABSTRACT_FROM' => 'lib/Plack/Middleware/Debug/Redis.pm',
    'LICENSE'       => 'perl',
    'PREREQ_PM'     => {
        'Redis'                    => '1.955',
        'Plack'                    => '0.9979',
        'Plack::Middleware::Debug' => 0,
        'strict'                   => 0,
    },
    'BUILD_REQUIRES' => {
        'Test::More'       => '0.98',
        'Test::Pod'        => '1.22',
        'Plack'            => '0.9979',
        'HTTP::Message'    => 0,
        'Test::MockObject' => 0,
    },
    'MIN_PERL_VERSION' => '5.10.1',
    'test' => {
        'TESTS' => 't/*.t',
    },
    'clean' => {
        'FILES' => '*.tar.gz',
    },
    $ExtUtils::MakeMaker::VERSION >= 6.50 ? (%$meta_merge) : (),
);

WriteMakefile(%options);


lib/Plack/Middleware/Debug/Redis.pm
package Plack::Middleware::Debug::Redis;

use strict;
use warnings;

our $VERSION = '0.02';


1; # End of Plack::Middleware::Debug::Redis
__END__

=head1 NAME

Plack::Middleware::Debug::Redis - Extend Plack::Middleware::Debug with Redis panels

=head1 VERSION

version 0.02

=head1 SYNOPSIS

    # inside your psgi app
    use Plack::Builder;

    my $app = sub {[
        200,
        [ 'Content-Type' => 'text/html' ],
        [ '<html><body>OK</body></html>' ]
    ]};
    my $redis_host = 'redi.example.com:6379';

    builder {
        mount '/' => builder {
            enable 'Debug',
                panels => [
                    [ 'Redis::Info', server => $redis_host ],
                    [ 'Redis::Keys', server => $redis_host, db => 3 ],
                ];
            $app;
        };
    };

=head1 DESCRIPTION

This distribution extends Plack::Middleware::Debug with some Redis panels. At the moment, the following panels
available:

=head1 PANELS

=head2 Redis::Info

Diplay panel with generic Redis server information which is available by the command INFO.
See L<Plack::Middleware::Debug::Redis::Info> for additional information.

=head2 Redis::Keys

Diplay panel with keys Redis server information. See L<Plack::Middleware::Debug::Redis::Keys>
for additional information.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug::Redis::Info>

L<Plack::Middleware::Debug::Redis::Keys>

L<Plack::Middleware::Debug>

L<Redis>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Plack/Middleware/Debug/Redis/Info.pm
package Plack::Middleware::Debug::Redis::Info;

use strict;
use warnings;
use v5.10.1;
use Redis 1.955;
use parent 'Plack::Middleware::Debug::Base';
use Plack::Util::Accessor qw/server password redis_handle/;

our $VERSION = '0.01';

sub prepare_app {
    my $self = shift;

    $self->server('localhost:6379') unless defined $self->server;

    my @opts = (
        server    => $self->server,
        reconnect => 60,
        encoding  => undef,
        debug     => 0,
    );
    push @opts, (password => $self->password) if $self->password;

    $self->redis_handle(Redis->new(@opts));
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Redis::Info');
    $panel->nav_title($panel->title);

    my $info = $self->redis_handle->info;

    # tweak db keys
    foreach my $db (grep { /^db\d{1,2}/ } keys %$info) {
        my $flatten = $self->flatten_db($db, $info->{$db});
        my @keys_flatten = keys %$flatten;
        @$info{@keys_flatten} = @$flatten{@keys_flatten};
        delete $info->{$db};
    }

    $panel->nav_subtitle('Version: ' . $info->{redis_version});

    return sub {
        $panel->content($self->render_hash($info));
    };
}

sub flatten_db {
    my ($self, $database, $value) = @_;

    my %flatten = ();

    %flatten = map {
        my @ary = split /=/;
        $database . '_' . $ary[0] => $ary[1];
    } split /,/, $value;

    \%flatten;
}


1; # End of Plack::Middleware::Debug::Redis::Info
__END__

=pod

=head1 NAME

Plack::Middleware::Debug::Redis::Info - Redis info debug panel

=head1 VERSION

version 0.01

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Redis::Info', server => 'redis.example.com:6379' ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Redis::Info extends Plack::Middleware::Debug by adding redis server info debug panel.
Panel displays data which available through INFO command issued in redis-cli. Before displaying info some tweaks
were processed. Normally INFO command shows total and expires keys in one line such as

    db0 => 'keys=167,expires=145',
    db1 => 'keys=75,expires=0',

This module turn in to

    db0_expires => '145',
    db0_keys    => '167',
    db1_expires => '0',
    db1_keys    => '75',

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 server

Hostname and port of redis server instance. Default value is 'localhost:6379'.

=head2 password

Password to authenticate on redis server instance in case of enabled redis' option B<requirepass>.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug>

L<Redis>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Plack/Middleware/Debug/Redis/Keys.pm
package Plack::Middleware::Debug::Redis::Keys;

use strict;
use warnings;
use v5.10.1;
use Redis 1.955;
use parent 'Plack::Middleware::Debug::Base';
use Plack::Util::Accessor qw/server password db redis_handle/;

our $VERSION = '0.01';

sub prepare_app {
    my $self = shift;

    $self->server('localhost:6379') unless defined $self->server;
    $self->db(0) unless defined $self->db;

    my @opts = (
        server    => $self->server,
        reconnect => 60,
        encoding  => undef,
        debug     => 0,
    );
    push @opts, (password => $self->password) if $self->password;

    $self->redis_handle(Redis->new(@opts));
}

sub run {
    my ($self, $env, $panel) = @_;

    $panel->title('Redis::Keys');
    $panel->nav_title($panel->title);

    return sub {
        my ($res) = @_;

        my ($keyz, $ktype, $klen);
        $self->redis_handle->select($self->db);
        my @keys = $self->redis_handle->keys('*');
        $panel->nav_subtitle('DB #' . $self->db . ' (' . scalar(@keys) . ')');

        for my $key (sort @keys) {
            $ktype = uc($self->redis_handle->type($key));

            given ($ktype) {
                when ('HASH')   { $klen = $self->redis_handle->hlen($key);   }
                when ('LIST')   { $klen = $self->redis_handle->llen($key);   }
                when ('STRING') { $klen = $self->redis_handle->strlen($key); }
                when ('ZSET')   { $klen = $self->redis_handle->zcard($key);  }
                when ('SET')    { $klen = $self->redis_handle->scard($key);  }
                default         { $klen = undef;                }
            }

            $keyz->{$key} = $ktype . ($klen ? ' (' . $klen . ')' : '');
        }

        $self->redis_handle->quit;
        $panel->content($self->render_hash($keyz));
    };
}

1; # End of Plack::Middleware::Debug::Redis::Keys
__END__

=pod

=head1 NAME

Plack::Middleware::Debug::Redis::Keys - Redis keys debug panel

=head1 SYNOPSIS

    # inside your psgi app
    enable 'Debug',
        panels => [
            [ 'Redis::Keys', server => 'redis.example.com:6379', db => 3 ],
        ];

=head1 DESCRIPTION

Plack::Middleware::Debug::Redis::Keys extends Plack::Middleware::Debug by adding redis server keys debug panel.
Panel displays available keys in the redis database and its type.

    coy:knows:pseudonoise:codes             STRING (9000)
    six:slimy:snails:sailed:silently        LIST (35)
    eleven:benevolent:elephants             HASH (17)
    two:tried:and:true:tridents             SET (101)
    tie:twine:to:three:tree:twigs           ZSET (66)

Also in brackets displays key-type specific data. For I<STRING> keys it's key length in bytes; for I<HASH> - number of fields
in a hash; for I<LIST> - length of a list (number of items); for I<SET> and I<ZSET> - number of members in a set.

This panel might be added several times for different databases. Just add it again to Plack Debug panels and provide another
database number.

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware::Debug>

=head2 run

See L<Plack::Middleware::Debug>

=head2 server

Hostname and port of redis server instance. Default value is 'localhost:6379'.

=head2 password

Password to authenticate on redis server instance in case of enabled redis' option B<requirepass>.

=head2 db

Redis database number to get statistic for keys. Default value is 0.

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Debug-Redis/issues>

=head1 SEE ALSO

L<Plack::Middleware::Debug>

L<Redis>

=head1 AUTHOR

Anton Gerasimov, E<lt>chim@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Anton Gerasimov

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
